Allow the cobra.mit.access.lookupBy* to have parameters of the query cus...

giovannimeo opened the pull request:
...tomized

- The cobra.mit.access.lookupBy* are useful routines, but lack teh
possibility to define filters, so providing a way for the API user to
supply them.

mtimm commented:
Thank you very much for the pull request.

Can you please address the new issues identified by landscape.io?

https://landscape.io/diff/117116

mtimm commented:
Also you can run the same tests locally to verify that you didn't introduce anything new using prospector:

https://github.com/datacenter/cobra/wiki/localLandscapeIo

mtimm commented on cobra/mit/access.py:
```diff
@@ -101,10 +101,11 @@ def lookupByDn(self, dnStrOrDn):
         return mos[0] if mos else None
 
-    def lookupByClass(self, classNames, parentDn=None, propFilter=None):
+    def lookupByClass(self, classNames, parentDn=None, propFilter=None, **kwargs):
```
Please remove propFilter from the arguments here as it is no longer used.  Also please update the docs for this method to no longer include propFilter but instead to include kwargs.

mtimm commented on cobra/mit/access.py:
```diff
@@ -125,12 +126,15 @@ def lookupByClass(self, classNames, parentDn=None, propFilter=None):
             mos = self.query(classQuery)
         return mos
+
+        def __setQueryParams(self, query, queryParams):
```
Please change setQueryParams to be a staticmethod and remove self from the arguments.

mtimm replied:
landscape.io is going to throw a warning on this method because of a lack of documentation.  Previously we were not very strict on this point for internal methods.  After thinking about it some more, it makes sense to document them, if anything for our own sanity.  So please add a doc string for this.  Even something as simple as:

```
"""Set the query parameters for the query.

An internal method to set parameters on a query.

Args:
  query (cobra.mit.request.AbstractQuery): The query object.
  queryParams (dict): The query parameters to set
 """
```

mtimm replied:
Also note that the indentation here is not correct.  It is indented 8 spaces but it should be 4.

mtimm commented on cobra/mit/access.py:
```diff
@@ -125,12 +126,15 @@ def lookupByClass(self, classNames, parentDn=None, propFilter=None):
         return mos
+
+        def __setQueryParams(self, query, queryParams):
+            for param, value in queryParams.iteritems():
```
iteritems is not Python 3 compatible.  Please change this to use one of these two idioms:

```
from future.utils import iteritems
...
for param, value in queryParams.iteritems():
```

Or:

```
from future.utils import viewitems
...
for param, value in queryParams.viewitems():
```

viewitems is more future proof as it is what python3 uses and eventually (many years to come I know) python2 will go away.

giovannimeo commented:
Thx Tim, i should have looked at the landscape.io before submitting the patch. I have addressed all the comments and tested. Will submit a new patch again

giovannimeo commented:
Tim, thanks for the comments. I have further updated the code following your comments. For the kwargs i used the example in:
http://sphinxcontrib-napoleon.readthedocs.org/en/latest/example_google.html

i have verified that pydoc represent it fine. sphinx is not documentating the functions as far as i can see from going in docs directory and doing 'make html'

mtimm commented:
Odd, the lookupByClass and lookupByDn should be documented but when I tried to build the docs locally with your changes they didn't seem to get updated.  If that continues to be a problem we can open an issue on it and address it later.  I will merge this now.  Thank you for contributing!

giovannimeo commented:
Tim, i have verified that 'pydoc cobra' launched from the toplevel documents the functions properly in cobra/mit/access.py, also after a clean rebuild i do see the **kwargs also being documented properly in sphinx.

giovannimeo commented:
Shinx created fine with kwargs
https://gist.github.com/giovannimeo/80d3d978489737d62a08#file-kwargs-sphinx-png
